source/common/upstream/upstream_impl.cc

```diff
@@ -67,7 +67,7 @@ ClusterInfoImpl::ClusterInfoImpl(const Json::Object& config, Runtime::Loader& ru
   ssl_ctx_ = nullptr;
   if (config.hasObject("ssl_context")) {
     Ssl::ContextConfigImpl context_config(*config.getObject("ssl_context"));
-    ssl_ctx_ = &ssl_context_manager.createSslClientContext(stats_scope_, context_config);
+    ssl_ctx_ = ssl_context_manager.createSslClientContext(stats_scope_, context_config);
```

Comment:
Inside the destructor of ContextImpl. 

FWIW, I'm not crazy about this solution, but I can't think of a better way. The caller is the only owner of the context, and the manager has a weak reference. It's cleaned up in destructor.